pkg/custom-provider/query_builder.go

```diff
@@ -0,0 +1,205 @@
+	//e.g. "namespace=my-namespace"
+	//e.g. "some_label=some-value"
+	var exprs []string
+
+	//Contains the list of label values we're targeting, by namespace.
+	//e.g. "some_label" => ["value-one", "value-two"]
+	valuesByName := map[string][]string{}
+
+	//Convert our query parts into template arguments.
+	for _, qPart := range queryParts {
+		//Be resilient against bad inputs.
+		//We obviously can't generate label filters for these cases.
+		if qPart.labelName == "" {
+			return nil, nil, NewLabelNotSpecifiedError()
+		}
+
+		if !n.operatorIsSupported(qPart.operator) {
+			return nil, nil, NewOperatorNotSupportedByPrometheusError()
+		}
+
+		matcher, err := n.selectMatcher(qPart.operator, qPart.values)
+
+		if err != nil {
+			return nil, nil, err
+		}
+
+		targetValue, err := n.selectTargetValue(qPart.operator, qPart.values)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		expression := matcher(qPart.labelName, targetValue)
+		exprs = append(exprs, expression)
+		valuesByName[qPart.labelName] = qPart.values
+	}
+
+	return exprs, valuesByName, nil
+}
+
+func (n *queryBuilder) selectMatcher(operator selection.Operator, values []string) (func(string, string) string, error) {
+
+	numValues := len(values)
+	if numValues == 0 {
+		switch operator {
+		case selection.Exists:
+			return prom.LabelMatches, nil
+		case selection.DoesNotExist:
+			return prom.LabelNotMatches, nil
+		case selection.Equals, selection.DoubleEquals, selection.NotEquals, selection.In, selection.NotIn:
+			return nil, NewOperatorRequiresValuesError()
+		}
+	} else if numValues == 1 {
+		switch operator {
+		case selection.Equals, selection.DoubleEquals:
+			return prom.LabelEq, nil
+		case selection.NotEquals:
+			return prom.LabelNeq, nil
+		case selection.In, selection.Exists:
+			return prom.LabelMatches, nil
+		case selection.DoesNotExist, selection.NotIn:
+			return prom.LabelNotMatches, nil
+		}
+	} else {
+		//Since labels can only have one value, providing multiple
+		//values results in a regex match, even if that's not what the user
+		//asked for.
+		switch operator {
+		case selection.Equals, selection.DoubleEquals, selection.In, selection.Exists:
+			return prom.LabelMatches, nil
+		case selection.NotEquals, selection.DoesNotExist, selection.NotIn:
+			return prom.LabelNotMatches, nil
+		}
+	}
+
+	return nil, errors.New("operator not supported by query builder")
+}
+
+func (n *queryBuilder) selectTargetValue(operator selection.Operator, values []string) (string, error) {
+	numValues := len(values)
+	if numValues == 0 {
+		switch operator {
+		case selection.Exists, selection.DoesNotExist:
+			//Regex for any non-empty string.
+			//When the operator is LabelNotMatches this will select series without the label
+			//or with the label but a value of "".
+			//When the operator is LabelMatches this will select series with the label
+			//whose value is NOT "".
+			return ".+", nil
```

Comment:
we don't actually need a regex here -- `Exists` is `name!=""` and `DoesNotExist` is `name==""`.